source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,416 @@
+    }
+  }
+
+  std::unordered_set<LbSubsetEntryPtr> subsets_modified;
+
+  for (const auto& host : hosts_added) {
+    for (const auto& keys : subset_keys_) {
+      // For each added host, for each subset key, attempt to extract the metadata
+      // corresponding to the key from the host.
+      SubsetMetadata kvs = extractSubsetMetadata(keys, *host);
+      if (!kvs.empty()) {
+        // The host has metadata for each key, find or create its subset.
+        LbSubsetEntryPtr entry = findOrCreateSubset(subsets_, kvs, 0);
+        ASSERT(entry != nullptr);
+        HostPredicate predicate =
+            std::bind(&SubsetLoadBalancer::hostMatches, this, kvs, std::placeholders::_1);
+        if (!entry->initialized()) {
+          entry->initLoadBalancer(*this, predicate);
+          if (entry->active()) {
+            stats_.lb_subsets_active_.inc();
+          }
+          stats_.lb_subsets_created_.inc();
+        } else if (subsets_modified.find(entry) == subsets_modified.end()) {
+          entry->host_subset_->update(hosts_added, hosts_removed, predicate);
+          subsets_modified.emplace(entry);
+        }
+      }
+    }
+  }
+
+  for (const auto& host : hosts_removed) {
+    for (const auto& keys : subset_keys_) {
+      // For each removed host, for each subset key, attempt to extract the metadata
+      // corresponding to the key from the host.
+      SubsetMetadata kvs = extractSubsetMetadata(keys, *host);
+      if (!kvs.empty()) {
+        // The host has metadata for each key, find its subset.
+        LbSubsetEntryPtr entry = findOrCreateSubset(subsets_, kvs, 0);
+        ASSERT(entry != nullptr);
+        if (!entry->initialized()) {
+          // Not found or not yet initialized: ignore this removed host since it
+          // was never part of the set.
+          continue;
+        }
+
+        if (subsets_modified.find(entry) == subsets_modified.end()) {
+          HostPredicate predicate =
+              std::bind(&SubsetLoadBalancer::hostMatches, this, kvs, std::placeholders::_1);
+          entry->host_subset_->update(hosts_added, hosts_removed, predicate);
+          subsets_modified.emplace(entry);
+        }
+      }
+    }
+  }
+
+  for (const auto& entry : subsets_modified) {
+    if (!entry->active()) {
+      stats_.lb_subsets_active_.dec();
+      stats_.lb_subsets_removed_.inc();
+    }
+  }
+}
+
+bool SubsetLoadBalancer::hostMatchesDefaultSubset(const Host& host) {
+  const envoy::api::v2::Metadata& host_metadata = host.metadata();
+
+  for (const auto& it : default_subset_.fields()) {
+    const ProtobufWkt::Value& host_value = Config::Metadata::metadataValue(
+        host_metadata, Config::MetadataFilters::get().ENVOY_LB, it.first);
+
+    if (!ValueUtil::equal(host_value, it.second)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool SubsetLoadBalancer::hostMatches(const SubsetMetadata& kvs, const Host& host) {
```

Comment:
Could this be a template function and merge with `hostMatchesDefaultSubset` ? Feel free to ignore if not worth it.

Reply:
Write a template function for this is easy. But it seems that the use of `std::bind` makes it so that I have to explicitly call out the specialization, which is pretty hard to read.

For example (post fix-format):

```
predicate = std::bind(&SubsetLoadBalancer::hostMatchesMetadata<
                          Protobuf::Map<ProtobufTypes::String, ProtobufWkt::Value>>,
                      this, std::placeholders::_1, default_subset_.fields());
```

I'm inclined to leave it as is, but I'll keep the change stashed if you rather the template function.

Reply:
Sure just drop it.

Reply:
I had the same reaction, but don't see a better solution.